plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange.hs

```diff
@@ -0,0 +1,107 @@
+                                                          TextDocumentIdentifier (TextDocumentIdentifier))
+import           Prelude                                 hiding (span)
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId)
+    { pluginHandlers = mkPluginHandler STextDocumentSelectionRange selectionRangeHandler
+    }
+
+selectionRangeHandler :: IdeState -> PluginId -> SelectionRangeParams -> LspM c (Either ResponseError (List SelectionRange))
+selectionRangeHandler ide _ SelectionRangeParams{..} = do
+    let (TextDocumentIdentifier uri) = _textDocument
+    let filePathMaybe = toNormalizedFilePath' <$> uriToFilePath' uri
+    case filePathMaybe of
+        Nothing -> pure . Right . List $ []
+        Just filePath -> liftIO $ do
+            let (List positions) = _positions
+            selectionRanges <- runIdeAction "SelectionRange" (shakeExtras ide) $ getSelectionRanges filePath positions
+            pure . Right . List $ selectionRanges
+
+getSelectionRanges :: NormalizedFilePath -> [Position] -> IdeAction [SelectionRange]
+getSelectionRanges file positions = fmap (fromMaybe []) <$> runMaybeT $ do
+    (HAR{hieAst, refMap}, positionMapping) <- useE GetHieAst file
+    positions' <- MaybeT . pure $ traverse (fromCurrentPosition positionMapping) positions
+    ast <- MaybeT . pure $ getAsts hieAst Map.!? (coerce. mkFastString . fromNormalizedFilePath) file
+    let ast' = runReader (preProcessAST ast) (PreProcessEnv refMap)
+    MaybeT . pure . traverse (toCurrentSelectionRange positionMapping) $
+        findSelectionRangesByPositions (astPathsLeafToRoot ast') positions'
+
+-- | Like 'toCurrentPosition', but works on 'SelectionRange'
+toCurrentSelectionRange :: PositionMapping -> SelectionRange -> Maybe SelectionRange
+toCurrentSelectionRange positionMapping SelectionRange{..} = do
+    newRange <- toCurrentRange positionMapping _range
+    pure $ SelectionRange {
+        _range = newRange,
+        _parent = _parent >>= toCurrentSelectionRange positionMapping
+    }
+
+-- | Build all paths from ast leaf to root
+astPathsLeafToRoot :: HieAST a -> [SelectionRange]
+astPathsLeafToRoot = mapMaybe (spansToSelectionRange . simplifySpans) . go [[]]
+  where
+    go acc (Node _ span [])       = fmap (span:) acc
+    go acc (Node _ span children) = concatMap (go (fmap (span:) acc)) children
+
+spansToSelectionRange :: [Span] -> Maybe SelectionRange
+spansToSelectionRange [] = Nothing
+spansToSelectionRange (span:spans) = Just $
+    SelectionRange {_range = realSrcSpanToRange span, _parent = spansToSelectionRange spans}
+
+simplifySpans :: [Span] -> [Span]
+simplifySpans = foldr go []
+  where
+    go x []     = [x]
+    go x (y:ys) = if x == y then y:ys else x:y:ys
+
+findSelectionRangesByPositions :: [SelectionRange] -> [Position] -> [SelectionRange]
+findSelectionRangesByPositions selectionRanges = fmap findByPosition
```

Comment:
I'd maybe just define the un-fmaped version and fmap it at the use site, but :shrug:

Reply:
This function filters the `[SelectionRange]`, preserving all `SelectionRange`s containing at least one of the `Positions`. (will add to comment later)
So, at first glance, I think there might exist better algorithm for this, at least better than the currently implemented simple traversal. That's why I keep it as a standalone function, rather than moving fmap to the caller site.

Reply:
@michaelpj Thanks for reading it so carefully. You got it right, the previous description is not accurate. I added more detailed explaination in the comment.

The performance concern makes sense somehow. If someone selects hundreds of positions in a batch edit, for a file containing tens of thousands of leaf nodes, it might become slow quickly. And will block the user's editing.

It is probably too early to optimize for things like that, for the very first version of this plugin. But at least for the possible optimizations I spotted during writing code, I would like to leave proper space for them. i.e. making the related code grouped together.

Reply:
Hm yes, I hadn't even thought about performance. In theory this is O(n*m) for the two input lists, but TBH I agree that they're almost certainly small and it doesn't matter.